### PR TITLE
Adds support for ctrl-tab and ctrl-shift-tab

### LINKED ISSUE #768: [Feature Request]: Switch tabs with CTRL+ TAB [$20]
It'd be nice to be able to switch open tabs the same way we can for Firefox: CTRL+TAB. and CTRL+SHIFT+TAB for reverse switch.

<bountysource-plugin>

---
There is a **[$20 open bounty](https://www.bountysource.com/issues/44432568-feature-request-switch-tabs-with-ctrl-tab?utm_campaign=plugin&utm_content=tracker%2F651521&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F651521&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by snaiperskaya96**
Fixes #768

### CODE REVIEW

**Comment by lukefromdc** [comment]
How do I test this, what are these to be used for?

**Comment by snaiperskaya96** [comment]
Hi,
please have a look at the ticket referenced on the first comment.
To test it compile it as usual and ctrl + tab/ctrl + shift + tab to switch tabs.

```
git clone https://github.com/snaiperskaya96/caja.git 
cd caja
git checkout add-support-for-ctrl-tab-and-ctrl-shift-tab
./autogen.sh
make
sudo make install
```

**Comment by sc0w** [comment]
I don't know why, but it doesn't work here, can someone else test it?

**Comment by lukefromdc** [comment]
@sc0w, what GTK version are you using? Wonder if that's the difference?

**Comment by snaiperskaya96** [comment]
Please make sure that there's no another caja process running. You could also try to restart

**Comment by sc0w** [comment]
@lukefromdc 

> what GTK version are you using? Wonder if that's the difference?

gtk 3.22.18 with the system updated in debian testing

@snaiperskaya96 

> Please make sure that there's no another caja process running. You could also try to restart

yes, I did, I restarted to make sure

**Comment by raveit65** [comment]
Hmm,  old keybindings are defined here https://github.com/mate-desktop/caja/blob/master/src/caja-navigation-window-menus.c#L868
But where are the new ones in you commit?

**Comment by snaiperskaya96** [comment]
@raveit65 In GTK you cannot define CTRL+SHIFT+TAB shortcut using an accelerator like normal bindings. So you have to intercept the key event and go ahead from there.
See `src/caja-navigation-window-menus.c`.

**Comment by snaiperskaya96** [comment]
Btw it's probably the same problem as Pluma (https://github.com/mate-desktop/pluma/pull/243) , these bindings work with the Mac keyboard but not with standard ones. Will edit tonight

**Comment by raveit65** [comment]
Not releated to your PR...
offtopic on
I am really missing the F6 key for switching panes in slit-view
https://github.com/mate-desktop/caja/blob/master/src/caja-navigation-window-menus.c#L851
Am i the only one where this doesn't work?
offtopic off

**Comment by raveit65** [comment]
ok, forgot my last comment, now F6 is working.

**Comment by snaiperskaya96** [comment]
If you have a moment, can you try this last commit please?

**Comment by raveit65** [comment]
Last commit makes crtl+tab and crtl+shift+tab working, but old keybindings are still working.
I would really like to see only one working keybings, to avoid that users are able to learn 2 different ways of switching tabs ;-)

**Comment by snaiperskaya96** [comment]
I kept both ways as per this comment https://github.com/mate-desktop/caja/issues/768#issuecomment-320886393

**Comment by raveit65** [comment]
I do not agree
@monsta @lukefromdc @sc0w 
Your opinion please.

**Comment by alexarnaud** [comment]
Le 24/08/2017 à 11:11, raveit65 a écrit :
> I do not agree
> @monsta <https://github.com/monsta> @lukefromdc 
> <https://github.com/lukefromdc> @sc0w <https://github.com/sc0w>
> Your opinion please.

IMO, it's not an issue if we could use both keyboard shortcuts. For 
example it's the behavior on Firefox. People accustomed to using 
ctrL+tab could used it and other that prefer ctrl+pgup/pgdown will be 
also happy.
It could also avoid to have a settings for choosing between them.

Best regards.
-- 
Alex ARNAUD
Visual-Impairment Project Manager
Hypra - "Humanizing technology"

**Comment by raveit65** [comment]
- ctrl+pgup/pgdown can be used for another function

**Comment by snaiperskaya96** [comment]
Not sure if my opinion could worth anything (lol), but if it does i think
users could only benefit from having both the shortcut available.
Will make old and new users happy, so the first ones do not have to change
habit, and the second ones will find standardised shortcuts.
It's also good to have shortcuts on both, left and right sides, i
personally find myself using different shortcuts based on what is more
comfortable for that moment (for example pasting in the terminal via
ctrl+shift+v or via shift insert)

**Comment by raveit65** [comment]
I think a teacher at school will be crazy for that, to teach 3 options for the same thing ;-)
But If everybody like 3 options for the same action i am not against it.
Personal i won't merge it before a statement from @monsta

**Comment by snaiperskaya96** [comment]
>I think a teacher at school will be crazy for that, to teach 3 options for the same thing ;-)
That's 'cause of the lazy nature of students (...and teachers actually) 😄 
It's always good to have alternatives imo, especially because, again, some persons might be used to older shortcuts, and other persons just prefer newer ones

**Comment by raveit65** [comment]
Btw. same problem here as with pluma PR.
Using crtl+tab for notebook tabs breaks using crtl+tab for the primary toolbar.
We need a gsettings key for switching from old to new behaviour.
Believe me, breaking old behaviour is the last thing what a linux user like ;-)

File: src/caja-navigation-window-menus.c [hunk 1]
```diff
@@ -819,6 +820,22 @@ action_tab_change_action_activate_callback (GtkAction *action, gpointer user_dat
...
 
+static gboolean
+on_key_press (GtkWidget *widget, GdkEventKey *event, gpointer user_data)
+{
+    if (event->keyval == GDK_KEY_ISO_Left_Tab && event->state & GDK_CONTROL_MASK)
+    {
+        action_tabs_previous_callback(NULL, user_data);
+        return TRUE;
+    }
+    if (event->keyval == GDK_KEY_Tab && event->state & GDK_CONTROL_MASK)
```
Comment (mspeers) [inline src/caja-navigation-window-menus.c]: Should this be else if statement? One or the other not both of them. I'm new so this project?
Also like this personalty
 if (event->state & GDK_CONTROL_MASK)
{
     switch (event->keyal)
.....
}

**Comment by sc0w** [comment]
already done in https://github.com/mate-desktop/caja/pull/1037

**Pull request closed without merging**